Merge latest blp_dev changes onto the WIP upstream merge

tylerparsons opened the pull request:
All features introduced by https://github.com/bloomberg/cartodb/pull/229

acarrera42 commented on app/models/carto/user.rb:
```diff
@@ -222,15 +225,48 @@ def google_maps_client_id
+  # if not it get everything else but GMaps in any case GMaps and other groups can work together
+  # this may have change in the future but in any case this method provides a way to abstract what
+  # basemaps are active for the user
+  #
```
Seems this comment is no longer valid if you remove the google_maps_enabled? logic below

tylerparsons replied:
See the end of the function - google maps logic is kept but moved to the end.  I think this comment shows up as an addition to the merge branch because it may have been deleted in the upstream, but it's been in our repo for 3+ years so I kept it: https://github.com/bloomberg/cartodb/blame/blp_dev/app/models/carto/user.rb#L203

acarrera42 commented on app/models/carto/user.rb:
```diff
@@ -222,15 +225,48 @@ def google_maps_client_id
+  #
+  # Basemaps may also be filtered by profile attributes - this implementation respects such filters.
   def basemaps
-    (Cartodb.config[:basemaps] || []).select { |group| group != 'GMaps' || google_maps_enabled? }
```
Although we don't enable google maps this will lead to merge conflicts later.  Is there an issue with simply keeping google_maps_enabled to false?

tylerparsons replied:
Logic kept and moved to the end of the function.

acarrera42 replied:
The logic appears to have moved to a new function filter_basemaps rather than the stay in the pre-existing base maps function.  Functionally it's moot since Bloomberg doesn't use google maps base layers.  I'm simply mentioning it for merging upstream carto code in the future.

tylerparsons replied:
`filter_basemaps` is a helper function within the scope of `basemaps` - the google maps logic still appears in the last line of `basemaps`.

acarrera42 commented on app/models/user.rb:
```diff
@@ -1453,8 +1463,40 @@ def google_maps_client_id
+  #
+  # Basemaps may also be filtered by profile attributes - this implementation respects such filters.
   def basemaps
-    (Cartodb.config[:basemaps] || []).select { |group| group != 'GMaps' || google_maps_enabled? }
```
Same comment as for model/user.rb

acarrera42 commented on app/models/visualization/overlays.rb:
```diff
@@ -8,6 +8,8 @@ def initialize(visualization)
 
       def create_default_overlays
+        prof_attrs = @visualization.user && @visualization.user.profile_attributes
+        puts "create_default_overlays - prof_attrs: #{prof_attrs}"
```
Remove trace or place in log if useful

tylerparsons replied:
Will remove - didn't mean to include this.

gfiorav replied:
Rubocop will aid in preventing things like this too! Please check it out

acarrera42 commented on db/migrate/20170630090000_add_category_foreign_key.rb:
```diff
@@ -7,6 +7,6 @@
   down do
     execute "ALTER TABLE visualizations DROP CONSTRAINT visualizations_category_fkey;"
-    execute "ALTER TABLE visualizations DROP CONSTRAINT unique_id;"
+    execute "ALTER TABLE visualization_categories DROP CONSTRAINT unique_id;"
```
Should migration scripts be modified?  Shouldn't we simply add a new migration script to make any changes?

mbektash replied:
this one is a bug fix actually

tylerparsons replied:
Correct, and it is also a change to the down task, not to the up task for an applied migration script, so it should be modified.

acarrera42 commented on db/migrate/20171114183148_profiles.rb:
```diff
@@ -0,0 +1,16 @@
+Sequel.migration do
+
+  up do
```
carto has a migration helper to avoid Postgres lockups: lib/carto/db/migration_helper.rb

Since these migrations were already applied though it may not make sense to modify these, but consider using the helper going forward.

tylerparsons replied:
Good to know. I'll take a look in the future but in general the overhead of adding a column should be minimal

gfiorav replied:
The helper won't interfere with normal operations. It should be used always for safety!

No need in your installation though, just looking for consistency.

gfiorav commented on app/controllers/api/json/profile_attributes_controller.rb:
```diff
@@ -0,0 +1,9 @@
+  include CartoDB
+
+  def show
+    render json: current_user.nil? ? {} : current_user.profile_attributes
```
Minor: you can rewrite as:

```ruby
render json: current_user.try(:profile_attributes) || {}
```

gfiorav commented on app/controllers/api/json/profile_attributes_controller.rb:
```diff
@@ -0,0 +1,9 @@
+
+class Api::Json::ProfileAttributesController < Api::ApplicationController
```
We use the `carto/` paths/modules for new controllers

gfiorav commented on app/controllers/api/json/profile_attributes_controller.rb:
```diff
@@ -0,0 +1,9 @@
+
+class Api::Json::ProfileAttributesController < Api::ApplicationController
+  include CartoDB
```
Why is this needed?

tylerparsons replied:
`include CartoDB` might not be needed.

gfiorav commented on app/controllers/carto/api/visualizations_controller.rb:
```diff
@@ -355,12 +355,16 @@ def count
         type = params.fetch(:type, 'datasets')
         typeList = (type == 'datasets') ? "'table','remote'" : "'derived'"
         categoryType = (type == 'datasets') ? 1 : 2
+        samples_user_id = sample_maps_user.id if categoryType == 2 && sample_maps_user
```
Not strictly requesting a change, but adding a comment for readability:

It's not immediate to the reader that having a `samples_user_id` is tied to not only having `sample_maps_user` (which is the straightforward part) but to also `categoryType == 2` (not related to `sample_user` in any nature). Same happens above with the `samples` condition. 

With a software project this big and so many contributors, it's important to follow the "least surprise" principle, even at the expense of some extra lines :)

gfiorav commented on app/controllers/carto/api/visualizations_controller.rb:
```diff
@@ -355,12 +355,16 @@ def count
         typeList = (type == 'datasets') ? "'table','remote'" : "'derived'"
         categoryType = (type == 'datasets') ? 1 : 2
+        samples_user_id = sample_maps_user.id if categoryType == 2 && sample_maps_user
+        sample_type_counts = Object.new
```
Initializing a variable like this isn't to common in Ruby. If you must, you can initialize to `nil`, and then invoke it as

```ruby
sample_type_counts || default_value
```

and set it like
```ruby
sample_type_counts ||= "whatever overrite if not set"
```

gfiorav commented on app/controllers/carto/api/visualizations_controller.rb:
```diff
@@ -387,10 +391,29 @@ def count
+                    FROM visualizations AS v
+                    WHERE v.user_id=? AND v.type IN (#{typeList}) AND v.locked=true
+                  ) AS results"
+          args = [user_id, user_id]
```
It's a good practice to initialize args right before using them.. so it's a bit more readable. Also, if they're not going to be used elsewhere, maybe you can save the extra variable.

```ruby
args = [user_id, user_id]
sample_type_counts = Sequel::Model.db.fetch(query, *args)
```

or 

```ruby
sample_type_counts = Sequel::Model.db.fetch(query, *[user_id, user_id])
```

gfiorav commented on app/controllers/carto/api/visualizations_controller.rb:
```diff
@@ -387,10 +391,29 @@ def count
+                  ) AS results"
+          args = [user_id, user_id]
+
+          query += ") AS results2"
```
Why can't you write that in the original query?

```ruby
         query = "SELECT
                COUNT(*) AS all,
                SUM(CASE WHEN likes > 0 THEN 1 ELSE 0 END) AS liked
              FROM (
                SELECT results.*, (SELECT COUNT(*) FROM likes WHERE actor=? AND subject=results.id) AS likes FROM (
                  SELECT v.id, v.type, v.category, v.locked
                    FROM visualizations AS v
                    WHERE v.user_id=? AND v.type IN (#{typeList}) AND v.locked=true
                  ) AS results ) AS resutls2"
```

gfiorav commented on app/models/carto/profile.rb:
```diff
@@ -0,0 +1,15 @@
+
+    validates :name, presence: true
+
+    has_many :profiles_users, :dependent => :destroy
```
Rocket operators are mostly deprecated. You can use `Rubocop` to lint the file with our config and tell you about these things :)

Should read:

`has_many :profiles_users, dependent: :destroy`

gfiorav commented on app/models/carto/user.rb:
```diff
@@ -40,6 +40,9 @@ class Carto::User < ActiveRecord::Base
   has_many :layers, through: :layers_user, after_add: Proc.new { |user, layer| layer.set_default_order(user) }
 
+  has_many :profiles_user, dependent: :destroy, foreign_key: :user_id, inverse_of: :user
+  has_many :profiles, :through => :profiles_user
```
Same here:

```ruby
has_many :profiles, through: :profiles_users
```

gfiorav commented on db/migrate/20171114183148_profiles.rb:
```diff
@@ -0,0 +1,16 @@
+  up do
+    create_table :profiles do
+      column :id, "uuid", :default=>Sequel::LiteralString.new("uuid_generate_v4()"),
+             :null=>false, :primary_key=>true
```
Same rocket operartor thing here

tylerparsons replied:
I can change, but is it advisable to change an already applied migration script though?  Seems unnecessary to risk changing the behavior

gfiorav replied:
Same, please say that this code is already in prod from the start (maybe you said it and I forgot). Not much point to CRing now.

gfiorav commented on db/migrate/20171114183148_profiles.rb:
```diff
@@ -0,0 +1,16 @@
+
+  up do
+    create_table :profiles do
+      column :id, "uuid", :default=>Sequel::LiteralString.new("uuid_generate_v4()"),
```
Why not use the usual `uuid` generator?
 
```ruby
Uuid :id, primary_key: true, null: false, unique: false, default: 'uuid_generate_v4()'.lit
```

Looking at the latest migrations and following those conventions is best.

tylerparsons replied:
Good to know in the future, but same concern as in `db/migrate/20171114183148_profiles.rb` - is it a good idea to change already applied migration scripts?

gfiorav replied:
Reviews should happen before the code is deployed :)

Agreed that it's late now

tylerparsons replied:
We actually did review in 4 separate pulls on bloomberg/cartodb - I think you were just not added as a reviewer at the time.

gfiorav replied:
That's great! 

Then don't request my review on this massive PR please, or be explicit about *what* you want me to review. No point in me investing time in this if it's already reviewed. I'd love to be a reviewer, but you need to ping me before the code is merged in order for the review to be as helpful as possible.

gfiorav commented on db/migrate/20171114183347_add_lock_flags_to_maps.rb:
```diff
@@ -0,0 +1,10 @@
+Sequel.migration do
+
+  change do
```
Write explicit `def up` and `def down` methods, or this behavior will be subject to whatever the next sequel gem update brings. Explicit is always best :)

tylerparsons replied:
I did make sure to test this beforehand with `db:migrate:up` and `db:migrate:down`.  If you think Sequel would introduce breaking changes to this functionality, I can change it.

gfiorav replied:
Explicit is always safer

tylerparsons commented:
Thanks @gfiorav. I should have been clearer in my original PR description, sorry for any confusion and duplicated efforts.  All the features in this PR were reviewed separately on bloomberg/cartodb, however there was a good bit of refactoring in merging the production code so I wanted to get your opinion on whether these pieces still fit well into the merge banch. Should have elaborated on the context of https://github.com/bloomberg/cartodb/pull/229

tylerparsons commented:
@gfiorav  I addressed most of those comments, agreed that these will be good improvements to the even though the code is deployed already.  Regarding specs, for better or worse, these changes were rolled without them because the tests for existing models/controllers were broken.  We have backlog items to fix these and write tests for this functionality as part of our next deployment.

gfiorav commented:
Hey @tylerparsons,

Yes, don't worry! It was just a bit of miscommunication :)

I can't really tell the refactors from the final code since I'm not familiar with the original one you made. My guess is that nothing is broken, but please test it thoroughly.

Thanks

tylerparsons commented:
Yes, manual testing has been quite thorough (painfully so) so we will definitely get those auto tests working :)
